django/utils/encoding.py

```diff
@@ -196,20 +193,46 @@ def iri_to_uri(iri):
     return quote(force_bytes(iri), safe=b"/#%[]=:;$&()+,!?*@'~")
 
 
+_hexdig = '0123456789ABCDEFabcdef'
+_hextobyte = None
+
+
 def uri_to_iri(uri):
     """
     Converts a Uniform Resource Identifier(URI) into an Internationalized
     Resource Identifier(IRI).
 
-    This is the algorithm from section 3.2 of RFC 3987.
+    This is the algorithm from section 3.2 of RFC 3987, excluding step 4.
 
     Takes an URI in ASCII bytes (e.g. '/I%20%E2%99%A5%20Django/') and returns
     unicode containing the encoded result (e.g. '/I \xe2\x99\xa5 Django/').
     """
     if uri is None:
         return uri
     uri = force_bytes(uri)
-    iri = unquote_to_bytes(uri) if six.PY3 else unquote(uri)
+
+    bits = uri.split(b'%')
+    if len(bits) == 1:
+        iri = uri
+    else:
+        parts = [bits[0]]
+        append = parts.append
+        global _hextobyte
+        if _hextobyte is None:
+            _allowed_ascii = [45, 46, 95, 126] + list(range(65, 91)) + list(range(97, 123))
+            _hextobyte = {('%02x' % char).encode(): six.int2byte(char)
+                          for char in _allowed_ascii}
+            _hextobyte.update(((a + b).encode(), six.int2byte(int(a + b, 16)))
+                              for a in _hexdig[8:] for b in _hexdig)
```

Comment:
I don't see much point in using a lazily computed global variable here. Can't this be moved to the module level?